src/ast_util.py

```diff
@@ -166,6 +166,126 @@ def generic_visit(self, node: ca.Node) -> None:
+        def visit_Struct(self, node: ca.Struct) -> None:
+            nonlocal found
+            if node is struct:
+                found = True
+            self.generic_visit(node)
+
+    for i, node in enumerate(ast.ext):
+        StructVisitor().visit(node)
+        if found:
+            return i
+    return -1
+
+
+def copy_struct(ast: ca.FileAST, typemap: TypeMap, struct: ca.Struct) -> ca.Struct:
+    """Adds a new struct with a definition matching `struct` to the AST.
+    Returns this new struct.
+    """
+    new_struct = copy.deepcopy(struct)
+    base_name = struct.name if struct.name else struct.decls[0].name
+    new_struct.name = get_noncolliding_struct_name(typemap, base_name)
+
+    # Insert the new struct right above the base struct. This preserves any
+    # struct field dependencies.
+    insert_idx = ast_ext_struct_idx(ast, struct)
+    decl = ca.Decl(name=None, quals=[], align=[], storage=[], funcspec=[], type=new_struct, init=None, bitsize=None)
+    ast.ext = ast.ext[:insert_idx] + [decl] + ast.ext[insert_idx:]
+    return new_struct
+
+
+def struct_set_field_type(parent: ca.Struct, field_idx: int, new_type: ca.Struct) -> None:
+    struct_type = copy.deepcopy(new_type)
+    struct_type.decls = None
+    old_type_decl = parent.decls[field_idx].type
+    assert isinstance(old_type_decl, ca.TypeDecl), "can only modify the type of a TypeDecl field"
+    old_type_decl.type = struct_type
+
+
+def struct_substruct_idx(parent: ca.Struct, substruct: ca.Struct, typemap: TypeMap) -> int:
+    """Returns the index of the field in `parent` of type `substruct`.
+    """
+    for i, decl in enumerate(parent.decls):
+        if isinstance(decl.type, ca.TypeDecl):
+            decl_struct = resolve_struct_def(decl.type, typemap)
+            if decl_struct is substruct:
+                return i
+    return -1
+
+
+def struct_rename_field(field: ca.Decl, name: str) -> None:
+    field.name = name
+    class TypeDeclVisitor(ca.NodeVisitor):
+        def visit_TypeDecl(self, node: ca.TypeDecl) -> None:
+            node.declname = name
+
+    TypeDeclVisitor().visit(field)
+
+
+def struct_remove_field(struct: ca.Struct, field_idx: int) -> None:
+    struct.decls = struct.decls[:field_idx] + struct.decls[field_idx+1:]
+
+
+def count_chained_structrefs(node: ca.StructRef) -> int:
+    """Count the number of chained StructRef's that are required to reach the
+    field in `node`.
+
+    Examples:
+      - foo            --> 0
+      - bar.foo        --> 1
+      - baz.bar.foo    --> 2
+      - bar->foo       --> 0
+      - baz.bar[i].foo --> 0
+    """
+    chain = 0
+    class StructRefVisitor(ca.NodeVisitor):
+        def visit_StructRef(self, node: ca.StructRef) -> None:
+            nonlocal chain
+            if not isinstance(node.name, ca.StructRef) or node.type != ".":
+                chain = 0
+                return
+            self.generic_visit(node)
+            chain += 1
+
+    StructRefVisitor().visit(node)
```

Comment:
```suggestion
    while isinstance(node.name, ca.StructRef) and node.type == ".":
        chain += 1
        node = node.name
```
Using a visitor confused me for a second, makes it look like you're visiting an entire AST.

Reply:
Done